File: frontend/src/app/services/discount.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Discount } from '../models/discount';

import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class DiscountService {
  private apiUrl = `${environment.apiUrl}/discounts`;

  constructor(private http: HttpClient) { }

  getAllDiscounts(): Observable<Discount[]> {
    return this.http.get<Discount[]>(this.apiUrl);
  }

  getDiscountById(id: number): Observable<Discount> {
    return this.http.get<Discount>(`${this.apiUrl}/${id}`);
  }

  createDiscount(discount: Discount): Observable<Discount> {
    return this.http.post<Discount>(this.apiUrl, discount);
  }

  updateDiscount(id: number, discount: Discount): Observable<Discount> {
    return this.http.put<Discount>(`${this.apiUrl}/${id}`, discount);
  }

  deleteDiscount(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getActiveDiscounts(): Observable<Discount[]> {
    return this.http.get<Discount[]>(`${this.apiUrl}/active`);
  }

  validateCoupon(code: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/validate?code=${code}`, {});
  }
}

File: backend/src/main/java/com/opticalshop/service/DiscountService.java
package com.opticalshop.service;

import com.opticalshop.model.Discount;
import com.opticalshop.model.Product;

import java.util.List;
import java.util.Optional;

public interface DiscountService {
    List<Discount> getAllDiscounts();

    Optional<Discount> getDiscountById(Long id);

    Discount createDiscount(Discount discount);

    Discount updateDiscount(Long id, Discount discount);

    void deleteDiscount(Long id);

    List<Discount> getActiveDiscounts();

    List<Discount> getDiscountsForProduct(Product product);

    Optional<Discount> getDiscountByCode(String code);

    Discount validateCoupon(String code) throws Exception;
}
